File: chapters/pc02/pc0202/c/pc0202.c
/**
 * pc110202
 * uva10315
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

typedef struct {
	int rank;
	int tiebreak[5];
} HandValue;

int card_value(char c) {
	if (c >= '2' && c <= '9') {
		return c - '0';
	}
	if (c == 'T') return 10;
	if (c == 'J') return 11;
	if (c == 'Q') return 12;
	if (c == 'K') return 13;
	return 14;
}

int cmp_desc(const void *a, const void *b) {
	int x = *(const int *)a;
	int y = *(const int *)b;
	return y - x;
}

void evaluate_hand(char cards[5][3], HandValue *hv) {
	int values[5];
	char suits[5];
	int count[15] = {0};
	int sorted[5];
	int freq_vals[5];
	int freq_cnts[5];
	int freq_size = 0;

	for (int i = 0; i < 5; i++) {
		values[i] = card_value(cards[i][0]);
		suits[i] = cards[i][1];
		count[values[i]]++;
		sorted[i] = values[i];
	}
	qsort(sorted, 5, sizeof(int), cmp_desc);

	int flush = 1;
	for (int i = 1; i < 5; i++) {
		if (suits[i] != suits[0]) {
			flush = 0;
			break;
		}
	}

	int straight = 1;
	for (int i = 0; i < 4; i++) {
		if (sorted[i] - 1 != sorted[i + 1]) {
			straight = 0;
			break;
		}
	}
	int straight_high = sorted[0];

	for (int v = 14; v >= 2; v--) {
		if (count[v] > 0) {
			freq_vals[freq_size] = v;
			freq_cnts[freq_size] = count[v];
			freq_size++;
		}
	}

	for (int i = 0; i < freq_size; i++) {
		for (int j = i + 1; j < freq_size; j++) {
			if (freq_cnts[j] > freq_cnts[i] ||
			   (freq_cnts[j] == freq_cnts[i] && freq_vals[j] > freq_vals[i])) {
				int tv = freq_vals[i];
				freq_vals[i] = freq_vals[j];
				freq_vals[j] = tv;
				int tc = freq_cnts[i];
				freq_cnts[i] = freq_cnts[j];
				freq_cnts[j] = tc;
			}
		}
	}

	memset(hv->tiebreak, 0, sizeof(hv->tiebreak));

	if (straight && flush) {
		hv->rank = 8;
		hv->tiebreak[0] = straight_high;
		return;
	}
	if (freq_cnts[0] == 4) {
		hv->rank = 7;
		hv->tiebreak[0] = freq_vals[0];
		hv->tiebreak[1] = freq_vals[1];
		return;
	}
	if (freq_cnts[0] == 3 && freq_cnts[1] == 2) {
		hv->rank = 6;
		hv->tiebreak[0] = freq_vals[0];
		return;
	}
	if (flush) {
		hv->rank = 5;
		for (int i = 0; i < 5; i++) hv->tiebreak[i] = sorted[i];
		return;
	}
	if (straight) {
		hv->rank = 4;
		hv->tiebreak[0] = straight_high;
		return;
	}
	if (freq_cnts[0] == 3) {
		hv->rank = 3;
		hv->tiebreak[0] = freq_vals[0];
		int k = 1;
		for (int i = 1; i < freq_size; i++) hv->tiebreak[k++] = freq_vals[i];
		return;
	}
	if (freq_cnts[0] == 2 && freq_cnts[1] == 2) {
		hv->rank = 2;
		hv->tiebreak[0] = freq_vals[0];
		hv->tiebreak[1] = freq_vals[1];
		hv->tiebreak[2] = freq_vals[2];
		return;
	}
	if (freq_cnts[0] == 2) {
		hv->rank = 1;
		hv->tiebreak[0] = freq_vals[0];
		int k = 1;
		for (int i = 1; i < freq_size; i++) hv->tiebreak[k++] = freq_vals[i];
		return;
	}

	hv->rank = 0;
	for (int i = 0; i < 5; i++) hv->tiebreak[i] = sorted[i];
}

int compare_hands(HandValue a, HandValue b) {
	if (a.rank != b.rank) {
		return (a.rank > b.rank) ? 1 : -1;
	}
	for (int i = 0; i < 5; i++) {
		if (a.tiebreak[i] != b.tiebreak[i]) {
			return (a.tiebreak[i] > b.tiebreak[i]) ? 1 : -1;
		}
	}
	return 0;
}

int main(void) {
	char black[5][3];
	char white[5][3];

	while (scanf("%2s %2s %2s %2s %2s %2s %2s %2s %2s %2s",
		black[0], black[1], black[2], black[3], black[4],
		white[0], white[1], white[2], white[3], white[4]) == 10) {
		HandValue hb, hw;
		evaluate_hand(black, &hb);
		evaluate_hand(white, &hw);

		int cmp = compare_hands(hb, hw);
		if (cmp > 0) {
			printf("Black wins.\n");
		} else if (cmp < 0) {
			printf("White wins.\n");
		} else {
			printf("Tie.\n");
		}
	}

	return 0;
}

File: chapters/pc02/pc0202/java/PC0202.java
/*
 * pc110202
 * uva10315
 */

package chapters.pc02.pc0202.java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.StringTokenizer;

public class PC0202 {

    static class HandValue {
        int rank;
        int[] tiebreak;

        HandValue(int rank, int[] tiebreak) {
            this.rank = rank;
            this.tiebreak = tiebreak;
        }
    }

    static int cardValue(char c) {
        if (c >= '2' && c <= '9') return c - '0';
        if (c == 'T') return 10;
        if (c == 'J') return 11;
        if (c == 'Q') return 12;
        if (c == 'K') return 13;
        return 14;
    }

    static HandValue evaluateHand(String[] cards) {
        List<Integer> values = new ArrayList<>();
        char[] suits = new char[5];
        Map<Integer, Integer> freq = new HashMap<>();

        for (int i = 0; i < 5; i++) {
            int v = cardValue(cards[i].charAt(0));
            values.add(v);
            suits[i] = cards[i].charAt(1);
            freq.put(v, freq.getOrDefault(v, 0) + 1);
        }

        values.sort(Collections.reverseOrder());

        boolean flush = true;
        for (int i = 1; i < 5; i++) {
            if (suits[i] != suits[0]) {
                flush = false;
                break;
            }
        }

        boolean straight = true;
        for (int i = 0; i < 4; i++) {
            if (values.get(i) - 1 != values.get(i + 1)) {
                straight = false;
                break;
            }
        }
        int straightHigh = values.get(0);

        List<int[]> grouped = new ArrayList<>();
        for (Map.Entry<Integer, Integer> e : freq.entrySet()) {
            grouped.add(new int[]{e.getValue(), e.getKey()});
        }
        grouped.sort((a, b) -> {
            if (a[0] != b[0]) return b[0] - a[0];
            return b[1] - a[1];
        });

        int[] tb = new int[5];

        if (straight && flush) {
            tb[0] = straightHigh;
            return new HandValue(8, tb);
        }
        if (grouped.get(0)[0] == 4) {
            tb[0] = grouped.get(0)[1];
            tb[1] = grouped.get(1)[1];
            return new HandValue(7, tb);
        }
        if (grouped.get(0)[0] == 3 && grouped.get(1)[0] == 2) {
            tb[0] = grouped.get(0)[1];
            return new HandValue(6, tb);
        }
        if (flush) {
            for (int i = 0; i < 5; i++) tb[i] = values.get(i);
            return new HandValue(5, tb);
        }
        if (straight) {
            tb[0] = straightHigh;
            return new HandValue(4, tb);
        }
        if (grouped.get(0)[0] == 3) {
            tb[0] = grouped.get(0)[1];
            tb[1] = grouped.get(1)[1];
            tb[2] = grouped.get(2)[1];
            return new HandValue(3, tb);
        }
        if (grouped.get(0)[0] == 2 && grouped.get(1)[0] == 2) {
            tb[0] = grouped.get(0)[1];
            tb[1] = grouped.get(1)[1];
            tb[2] = grouped.get(2)[1];
            return new HandValue(2, tb);
        }
        if (grouped.get(0)[0] == 2) {
            tb[0] = grouped.get(0)[1];
            tb[1] = grouped.get(1)[1];
            tb[2] = grouped.get(2)[1];
            tb[3] = grouped.get(3)[1];
            return new HandValue(1, tb);
        }

        for (int i = 0; i < 5; i++) tb[i] = values.get(i);
        return new HandValue(0, tb);
    }

    static int compareHands(HandValue a, HandValue b) {
        if (a.rank != b.rank) return a.rank > b.rank ? 1 : -1;
        for (int i = 0; i < 5; i++) {
            if (a.tiebreak[i] != b.tiebreak[i]) return a.tiebreak[i] > b.tiebreak[i] ? 1 : -1;
        }
        return 0;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder out = new StringBuilder();
        String line;

        while ((line = br.readLine()) != null) {
            if (line.trim().isEmpty()) {
                continue;
            }

            StringTokenizer st = new StringTokenizer(line);
            String[] black = new String[5];
            String[] white = new String[5];

            for (int i = 0; i < 5; i++) black[i] = st.nextToken();
            for (int i = 0; i < 5; i++) white[i] = st.nextToken();

            HandValue hb = evaluateHand(black);
            HandValue hw = evaluateHand(white);

            int cmp = compareHands(hb, hw);
            if (cmp > 0) out.append("Black wins.\n");
            else if (cmp < 0) out.append("White wins.\n");
            else out.append("Tie.\n");
        }

        System.out.print(out);
    }
}
